crates/oxc_linter/src/ast_util.rs

```diff
@@ -975,3 +975,149 @@ pub fn is_node_within_call_argument<'a>(
+            Some(expr_stmt.span)
+        } else {
+            None
+        }
+    }) else {
+        return false;
+    };
+
+    // Check if our node starts at the same position as the ExpressionStatement.
+    // If not, our replacement won't be at the start of the statement, so no ASI hazard.
+    // For example: `foo.bar(Array.from(x))` - the Array.from is not at statement start.
+    if node_span.start != expr_stmt_span.start {
+        return false;
+    }
+
+    // Now check if the character before the ExpressionStatement could cause ASI issues
+    if expr_stmt_span.start == 0 {
+        return false;
+    }
+
+    let source_text = ctx.source_text();
+
+    // Find the last meaningful character before the span, skipping whitespace and comments.
+    let last_char = find_last_meaningful_char(source_text, expr_stmt_span.start, ctx);
+
+    let Some(last_char) = last_char else {
+        return false;
+    };
+
+    // These characters, when followed by `[`, `(`, `/`, `+`, `-`, or `` ` ``,
+    // could cause the new code to be parsed as a continuation of the previous expression
+    matches!(last_char, ')' | ']' | '}' | '"' | '\'' | '`' | '+' | '-')
+        || last_char.is_alphanumeric()
+        || last_char == '_'
+        || last_char == '$'
+}
+
+/// Find the last meaningful (non-whitespace, non-comment) character before the given position.
+/// Uses semantic's comment tracking to skip entire comment spans efficiently.
+fn find_last_meaningful_char(source_text: &str, end_pos: u32, ctx: &LintContext) -> Option<char> {
+    let bytes = source_text.as_bytes();
+    let mut i = end_pos;
+
+    while i > 0 {
+        i -= 1;
+
+        let byte = bytes[i as usize];
+
+        // Skip ASCII whitespace (single byte)
+        if byte.is_ascii_whitespace() {
+            continue;
+        }
+
+        if let Some(comment) = ctx.semantic().get_comment_at(i) {
+            i = comment.span.start;
+            continue;
+        }
+
+        // Found a meaningful byte. Now we need to decode the character properly.
+        // For ASCII characters (byte < 128), we can directly convert.
+        // For multi-byte UTF-8 characters, we need to find the start of the character.
+        if byte.is_ascii() {
+            return Some(byte as char);
+        }
+
+        // Multi-byte UTF-8 character: find the start byte.
+        // UTF-8 continuation bytes have the pattern 10xxxxxx (0x80-0xBF).
+        // Start bytes have patterns 0xxxxxxx, 110xxxxx, 1110xxxx, or 11110xxx.
+        let mut char_start = i as usize;
+        while char_start > 0 && (bytes[char_start] & 0b1100_0000) == 0b1000_0000 {
+            char_start -= 1;
+        }
+
+        // Decode the character from the start position
+        if let Some(c) = source_text[char_start..].chars().next() {
+            return Some(c);
+        }
+
+        // Fallback: shouldn't happen with valid UTF-8, but be safe
+        return None;
```

Comment:
I would unwrap here. If logic is incorrect, we want a panic, not silently incorrectness.